brayns/Brayns.cpp

```diff
@@ -46,72 +44,44 @@ namespace brayns
 struct Brayns::Impl
 {
     Impl( int argc, const char **argv )
-         : _frameSize( 0, 0 )
-         , _sceneModified( true )
+         : _sceneModified( true )
     {
-        // Ignore errors from multiple calls to init from same process,
-        // aka from unit tests. Remove this once ospShutdown is provided.
-        try { ospInit( &argc, argv ); }
-        catch( const std::runtime_error& ) {}
-
         BRAYNS_INFO << "Parsing command line options" << std::endl;
         _parametersManager.reset( new ParametersManager( ));
         _parametersManager->parse( argc, argv );
         _parametersManager->print( );
 
-        _frameSize =
-            _parametersManager->getApplicationParameters( ).getWindowSize( );
-
-        BRAYNS_INFO << "Initialize renderers" << std::endl;
-        // Should be implemented with a plugin factory
-        _activeRenderer =
-            _parametersManager->getRenderingParameters().getRenderer();
-
-        const strings& renderers =
-            _parametersManager->getRenderingParameters().getRenderers();
-
-        Renderers renderersForScene;
-        for( std::string renderer: renderers )
-        {
-            _renderers[renderer].reset(
-                new OSPRayRenderer( renderer, *_parametersManager ));
-            renderersForScene.push_back( _renderers[renderer] );
-        }
-
-        BRAYNS_INFO << "Initialize scene" << std::endl;
-        _scene.reset( new OSPRayScene( renderersForScene,
-            _parametersManager->getSceneParameters(),
-            _parametersManager->getGeometryParameters()));
-
-        _scene->setMaterials( MT_DEFAULT, NB_MAX_MATERIALS );
+        // Get rendering engine
+        EngineFactory engineFactory( argc, argv, _parametersManager );
+        std::string engineName =
+            _parametersManager->getRenderingParameters( ).getEngine();
+        _engine = engineFactory.get( engineName );
+        if( !_engine )
+            throw std::runtime_error( "Unsupported engine: " + engineName );
 
         // set HDRI skybox if applicable
         const std::string& hdri =
             _parametersManager->getRenderingParameters().getHDRI();
+        ScenePtr scene = _engine->getScene();
         if( !hdri.empty() )
-            _scene->getMaterial(MATERIAL_SKYBOX)->setTexture(TT_DIFFUSE, hdri);
+            scene->getMaterial(MATERIAL_SKYBOX)->setTexture(TT_DIFFUSE, hdri);
 
         // Default sun light
         DirectionalLightPtr sunLight( new DirectionalLight(
             Vector3f( 0.f, 0.f, 1.f ), Vector3f( 1.f, 1.f, 1.f ), 1.f ));
-        _scene->addLight( sunLight );
+        scene->addLight( sunLight );
 
+        // Build geometry
         loadData( );
-        _scene->buildEnvironment( );
-        _scene->buildGeometry( );
-        _scene->commit( );
+        scene->buildEnvironment( );
+        scene->buildGeometry( );
+        scene->commit( );
```

Comment:
Well, not really. GPU engine such as Optix also need to transfer data from CPU to GPU memory and this is the job of the commit method. That will eventually also be useful for out-of-core rendering.